add wormhole docs (#196)
PureStake/moonbeam-docs-cn: builders/interoperability/protocols/wormhole.md

---
title: 通过Wormhole进行跨链操作
description: 学习如何使用Wormhole桥接资产、设置中继器以及将Moonbeam dApp连接到多个区块链上的资产和功能的其他方法。
---

# Wormhole Network

![Wormhole Moonbeam banner](/images/builders/integrations/bridges/wormhole/wormhole-banner.png)

## 概览

[Wormhole](https://wormhole.com/){target=_blank}是一种通过称为VAAs（Verifiable Action Approvals）的消息来为Web3验证和保护跨链通信的协议。Wormhole的基础设施能够使dApp用户通过一键操作与任何链上的任意资产或应用程序交互。有一个多签协议和19个签署[Guardians](https://book.wormhole.com/wormhole/5_guardianNetwork.html){target=_blank}赋能, Wormhole可以让dApps跨链传输任意消息.

> Wormhole由多个模块化的交换组件组成，这些组件可以独立使用并支持由多个团队构建的逐渐增加的可组合应用程序。在其协议之上构建xDapp允许快速进行跨链资产转移和跨链逻辑以提供最大的Web互操作性。Wormhole的架构包含签署Guardian网络、桥接智能合约和中继器。查看技术堆栈以获取更多信息。

Reviewer: ```suggestion
Wormhole由多个模块化的交换组件组成，这些组件可以独立使用并支持由多个团队构建的逐渐增加的可组合应用程序。在其协议之上构建xDapp允许快速进行跨链资产转移和跨链逻辑以提供最大的Web互操作性。Wormhole的架构包含签署Guardian网络、桥接智能合约和中继器。请查看技术栈图以获取更多信息。
```

![Wormhole Technology Stack diagram](/images/builders/integrations/bridges/wormhole/wormhole-1.png)

--8<-- 'text/disclaimers/third-party-content-intro.md'

## 开始操作 {: #getting-started }

> 开始使用Wormhole构建跨链应用程序可能所需资源：

Reviewer: ```suggestion
开始使用Wormhole构建跨链应用程序可能所需的资源：
```

- **[开发者文档](https://book.wormhole.com/introduction/introduction.html){target=_blank}** - 技术性指南
- **[门户网站](https://www.portalbridge.com/#/transfer){target=_blank}** - 跨链转移资产的桥接UI

## 合约 {: #contracts }

查看部署至Moonbeam的Wormhole合约列表，以及通过Wormhole连接至Moonbeam的网络。

- **主网合约** - [Moonbeam](https://book.wormhole.com/reference/contracts.html#mainnet){target=_blank}

- **测试网合约** - [Moonbase Alpha](https://book.wormhole.com/reference/contracts.html#testnet){target=_blank}

## 使用Relayer Engine设置专用中继器 {: #setting-up-a-specialized-relayer-with-the-relayer-engine }

在此部分中，您将部署一个基本的Wormhole互连智能合约并启动一个专用中继器用于发送跨链消息。

> 首先，您需要了解一些背景。 VAAs（Verifiable Action Approvals）是Wormhole版本的跨链消息验证格式。如果Wormhole的19名守护者有13名验证了一则特定消息，则该消息将获得批准并在其他链上接收。与Guardian（守护者）网络（充当Wormhole协议的验证人）相邻的是网络间谍。他们不做任何验证工作。相反，他们会监视守护者网络并作为一个页面以允许用户和应用查看哪些VAAs已获得批准。

Reviewer: ```suggestion
首先，您需要了解一些背景。 VAAs（Verifiable Action Approvals）是Wormhole版本的经过验证的跨链消息。如果Wormhole的19名守护者有13名验证了一则特定消息，则该消息将获得批准并在其他链上接收。与Guardian（守护者）网络（充当Wormhole协议的验证人）相邻的是网络间谍。他们不做任何验证工作。相反，他们会监视守护者网络并作为一个页面以允许用户和应用查看哪些VAAs已获得批准。
```

中继器的职责是为目标链的执行支付费用，且在许多通用中继器中，中继器又是由用户支付费用。Wormhole目前还没有该功能，所以Wormhole的架构需要dApp开发者创建并维护其自己的专用中继器。一个开发者如果想要合约调用者为目标链支付gas费用，则需要设计自己的系统。这可能看起来工作量很大，但这允许对消息处理的方式进行微调。举例而言，一个中继器可以同时发送相同的消息至多条链。

### 查看先决条件 {: #checking-prerequisites }

在开始跟随本教程操作之前，您需要准备以下内容：

- [安装MetaMask并连接至Moonbase Alpha](/tokens/connect/metamask/){target=_blank}
- [安装Docker](https://docs.docker.com/get-docker/){target=_blank}
- 准备一个账户，为该账户提供一定数量的`DEV` token.
   --8<-- 'text/faucet/faucet-list-item.md'
- 使用同一个账户，从您选择的Wormhole连接的EVM中提供原生资产至该账户。Faucet[如下方图表所示](#deploying-the-wormhole-contract-with-remix-on-moonbase-alpha)

### 使用Remix在Moonbase Alpha上部署Wormhole合约 {:deploying-the-wormhole-contract-with-remix-on-moonbase-alpha}

要发送跨链消息，在本教程中，你将需要部署和使用一个智能合约。每条连接至Wormhole的链都会有某种[Wormhole核心桥接](https://github.com/wormhole-foundation/wormhole/blob/dev.v2/ethereum/contracts/interfaces/IWormhole.sol){target=_blank}的实现，其功能为发布和验证VAAs。每一个核心桥接合约（每条链一个）的实现均由守护者网络上的守护者见证，这便是他们得知何时开始验证一条消息的方式。

不同于其他跨链合约，Wormhole并不提供用户可以继承并在其上构建的母合约。这是因为Wormhole的第一条链Solana，不像Solidity一样在其智能合约中提供特有的继承。为了让每条链上的设计体验相似，Wormhole让他们的Solidity开发者直接与EVM链上的Wormhole核心桥接智能合约交互。

> 本文中要部署的[智能合约](https://github.com/jboetticher/relayer-engine/blob/main/SimpleGeneralMessage.sol){target=_blank}储存在一个由Wormhole的Relayer Engine存储库分叉出来的Git存储库中。这将从一条链向另一条链发送字符串，并在通过Wormhole协议接收到字符串时进行存储。要部署脚本，将合约复制并粘贴至Remix或打开[Remix gist链接](https://remix.ethereum.org/?#gist=6aac8f954e245d6394f685af5d404b4b&optimize=false&runs=200&evmVersion=null&version=soljson-v0.8.17+commit.8df45f5f.js){target=_blank}。

Reviewer: ```suggestion
本文中要部署的[智能合约](https://github.com/jboetticher/relayer-engine/blob/main/SimpleGeneralMessage.sol){target=_blank}储存在一个由Wormhole的Relayer Engine存储库分叉出来的Git存储库中。这将从一条链向另一条链发送字符串，并在通过Wormhole协议接收到字符串时进行存储。要部署脚本，将合约复制并粘贴至Remix或打开[Remix gist链接](https://remix.ethereum.org/?#gist=6aac8f954e245d6394f685af5d404b4b){target=_blank}。
```

> 首先，该智能合约中的代码是基于[Wormhole的最佳开发文档](https://book.wormhole.com/technical/evm/bestPractices.html){target=_blank}，但某些方面（如安全性）已被简化。当您编写用于生产环境的智能合约时，**请检查文档以更好地了解其标准**。

Reviewer: ```suggestion
首先，该智能合约中的代码是基于[Wormhole的最佳实践开发文档](https://book.wormhole.com/technical/evm/bestPractices.html){target=_blank}，但某些方面（如安全性）已被简化。当您编写用于生产环境的智能合约时，请查阅文档以更好地了解其标准。需要明确的是，**不要在生产环境中使用以下智能合约**。
```

|                          Network & Faucet                           |             Core Bridge Address            | Wormhole Chain ID |
|:-------------------------------------------------------------------:|:------------------------------------------:|:-----------------:|
| [Polygon Mumbai](https://faucet.polygon.technology/){target=_blank} | 0x0CBE91CF822c73C2315FB05100C2F714765d5c20 |         5         |
|    [Avalanche Fuji](https://faucet.avax.network/){target=_blank}    | 0x7bbcE28e64B3F8b84d876Ab298393c38ad7aac4C |         6         |
|   [Fantom TestNet](https://faucet.fantom.network/){target=_blank}   | 0x1BB3B4119b7BA9dfad76B0545fb3F531383c3bB7 |         10        |
|         [Goerli](https://goerlifaucet.com/){target=_blank}          | 0x706abc4E45D419950511e474C7B9Ed348A4a716c |         2         |
| [Moonbase Alpha](https://docs.moonbeam.network/builders/get-started/networks/moonbase/#moonbase-alpha-faucet){target=_blank} | 0xa5B7D85a8f27dd7907dc8FdC21FA5657D5E2F901 | 16 |

1. 确保选择的合约为**SimpleGeneralMessage**
2. 点击箭头按钮打开部署菜单
3. 在**_CHAINID**输入框内输入相关chain ID
4. 在**WORMHOLE_CORE_BRIDGE_ADDRESS**输入框内输入相关核心桥接地址
5. 点击**transact**按钮以开始部署交易
6. 点击MetaMask中的**Confirm**按钮以开始部署

> 当您在Moonbase Alpha上成功部署合约后，请确保复制其地址并使用连接到Wormhole的任何[EVM测试网](https://layerzero.gitbook.io/docs/technical-reference/testnet/testnet-addresses){target=_blank}重复该流程，以确保它可以跨链发送消息。注意您需要在MetaMask更改网络以部署至正确的网络。

Reviewer: ```suggestion
当您在Moonbase Alpha上成功部署合约后，请确保复制其地址并使用连接到Wormhole的任何其他[EVM测试网](https://layerzero.gitbook.io/docs/technical-reference/testnet/testnet-addresses){target=_blank}重复该流程，以确保您可以跨链发送消息。注意您需要在MetaMask更改网络以部署至正确的网络。
```

### 将Moonbase Alpha的互连合约列入白名单 {:whitelisting-moonbase-alpha-connected-contract}

此时，您需要将相同的智能合约部署两次。 一个在Moonbase Alpha上，另一个在另外的EVM链上。

Wormhole建议在互连合约中包含一个白名单系统，您将在尝试发送跨链消息前用于`SimpleGeneralMessage`。

要添加白名单合约，您必须调用`addTrustedAddress(bytes32 sender, uint16 _chainId)`函数，这需要一个*bytes32*格式的地址和一个chain ID。您可以在[上述表格](#deploying-the-wormhole-contract-with-remix-on-moonbase-alpha)中以及[Wormhole的文档网站](https://book.wormhole.com/reference/contracts.html#testnet){target=_blank}中找到chain ID。

```javascript
function addTrustedAddress(bytes32 sender, uint16 _chainId) external {
    myTrustedContracts[sender][_chainId] = true;
}
```

> 注意`sender`参数是_bytes32_而非_地址_类型。Wormhole的VAAs以_bytes32_的形式提供发射器（源）地址，所以它们以_bytes32_的形式存储和检查。若要将_地址_类型转换为_bytes32_，您将需要再加上24个0。这是因为_地址_的值是20个bytes，小于_bytes32_的32个。每个byte有两个十六进制字节，所以：

Reviewer: ```suggestion
注意`sender`参数是一个_bytes32_类型，而非_address_类型。Wormhole的VAAs以_bytes32_的形式提供发射器（源）地址，所以它们以_bytes32_的形式存储和检查。若要将_address_类型转换为_bytes32_，您将需要再加上24个0。这是因为_address_的值是20个bytes，小于_bytes32_的32个。每个byte有两个十六进制字符，所以：
```

![Add trusted address](/images/builders/integrations/bridges/wormhole/wormhole-5.png)

在此部分中，您应该已经将两条链上的两笔交易发送到两个合约中的白名单地址。之后，您应该被允许在互连合约之间发送消息。

### 运行Wormhole守护者网络间谍 {: #running-wormhole-guardian-spy }

现在您将为Wormhole运行一个测试网中继器！本教程是基于Wormhole的[relayer-engine](https://github.com/wormhole-foundation/relayer-engine){target=_blank} Github存储库的，截至本文撰写时，该存储库已提交至[`dac6012`](https://github.com/wormhole-foundation/relayer-engine/tree/dac6012cc7ed9c3ca79d911b47f39bfe9dd76a23){target=_blank}。目前仍处于开发阶段，文件夹结构可能会发生比较大的变化。

克隆专门为与`SimpleGeneralMessage`交互的[relayer-engine分叉](https://github.com/jboetticher/relayer-engine){target=_blank}。运行该中继器需要[Docker](https://docs.docker.com/get-docker/){target=_blank}和[npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm){target=_blank}，所以请确保您已将其安装到您的设备上。

首先需要设置。使用npm包管理器通过命令行安装依赖项。

```
npm install
cd example-project
npm install
```

完成之后，查看不同的文件夹。有两种主要的文件夹：`relayer-engine`和`example-project`。`relayer-engine`包含帮助运行中继器的组件，而`example-project`文件夹包含特定于 `SimpleGeneralMessage`智能合约的插件脚本和配置文件。此外，在根目录中还有一个`README.md`文件，其将包含有关中继器以及如何设置中继器的更多信息。

但是在深入了解如何运行任何东西或任何插件脚本如何工作之前，您首先需要了解中继器的不同组件以及中继器的作用。

> 中继器过滤并接收来自守护者网络的VAA，并对此进行处理。在此情况下，收集器会过滤来自您部署的互连合约通过批准的消息，然后解析该VAA，确定其目的地，最终尝试在目的地执行一个叫做`processMyMessage(bytes32 VAA)`的函数。您需要了解的是来自其他参与者的中继器可以收到这个VAA且其他中继器可以通过任何方式执行任何VAA。

Reviewer: ```suggestion
中继器过滤并接收来自守护者网络的VAA，并对此进行处理。在此情况下，中继器会过滤来自您部署的互连合约的守护者批准的消息，然后解析该VAA，确定其目的地，最终尝试在目的地执行一个叫做`processMyMessage(bytes32 VAA)`的函数。您需要了解的是来自其他参与者的中继器可以收到这个VAA且其他中继器可以以他们认为合适的任何方式执行任何VAA。
```

从技术角度来看，该中继器的实现分为四个部分。

> 1. 一个监视所有VAA的Wormhole守护者网络的非验证间谍节点

Reviewer: ```suggestion
1. 一个为所有VAA监视Wormhole守护者网络的非验证间谍节点
```

2. 一个称为监听器的组件，其收集任何间谍节点的输出，过滤出与中继器相关信息，然后将其打包至工作流对象中
3. 存储监听器输出的工作流对象的Redis数据库
4. 一个称为执行者的组件，从数据库中弹出工作流并以某种方式处理（在此情况下，在目标链上发送交易）

> Docker容器负责监视间谍节点，该Docker容器很容易启动。`relayer-engine`包存储于一个类似于命名的文件夹中，其中包含用于监听器和数据库的大部分代码。执行器的大部分逻辑都将依赖于开发者编写的插件（即从存储库克隆的插件），但大多样板代码仍由`relayer-engine`包处理。

Reviewer: ```suggestion
Docker容器负责监视间谍节点，该Docker容器很容易启动。`relayer-engine`包存储于一个采用类似命名的文件夹中，其中包含用于监听器和数据库的大部分代码。执行器的大部分逻辑都将依赖于开发者编写的插件（即从存储库克隆的插件），但大多样板代码仍由`relayer-engine`包处理。
```

泄露密钥可能导致资金流失，因此请妥善保管您的密钥。虽然`executor.json`在存储库中被git忽略，请确保您在测试网使用的钱包中没有任何主网资金。

```javascript
{
   "privateKeys": {
       "16": [
           "0x4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d"
       ],
       "2": [
           "0x4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d"
       ],
       "5": [
           "0x4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d"
       ],
       "6": [
           "0x4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d"
       ],
       "10": [
           "0x4f3edf983ac636a65a842ce7c78d9aa706d3b113bce9c46f30d7d21715b23b1d"
       ],

   }
}
```

如果您正在使用一条并未在上方EVM测试网列表中列出的链，您将需要添加您自己的数组。该数组的密钥应该在另一个您之前决定部署的EVM的Wormhole chain ID。例如，如果您在Fantom TestNet上部署，您将添加以下对象，因为Fantom TestNet的Wormhole chain ID是10。

```javascript
"10": [
    "YOUR PRIVATE KEY"
]
```

现在执行器的钱包已处理完成，我们看看执行器代码本身，其存在于`example-project/plugins/simplegeneralmessage_plugin/src/plugin.ts` 文件中。若您没有跟随操作，可以在[其GitHub存储库](https://github.com/jboetticher/relayer-engine/blob/main/example-project/plugins/simplegeneralmessage_plugin/src/plugin.ts){target=_blank}获取整个文件。

> `handleWorkflow(workflow, providers, execute)`函数是逻辑中心，不过其下也有一些辅助函数。这是`relayer-engine`包在Redis数据库中有工作流要被使用时调用的函数。注意注入函数的三个参数：`workflow`、`providers`和`execute`。

Reviewer: ```suggestion
所有逻辑都发生在`handleWorkflow(workflow, providers, execute)`函数里，不过其下也有一些辅助函数。这是`relayer-engine`包在Redis数据库中有工作流要被使用时调用的函数。注意注入函数的三个参数：`workflow`、`providers`和`execute`。
```

- `workflow`对象在监听器组件的`consumeEvent(vaa, stagingArea)`函数执行时提供存放在数据库的数据。在这种境况下，仅有VAA以及其被接收的时间会被存放至数据库，其被存放在本地`payload`变量中
- `providers`对象注入Ethers以及其他链提供商，这对于查询链上数据或进行其他区块链相关操作可能有帮助。如之前所述，目前被该包支持的提供商仅有Solana和EVM。`providers`对象不被用于实现

> - `execute`对象中目前有两个函数：`onEVM(options)`和`onSolana(options)`。这些函数需要一个Wormhole chain ID和一个有钱包注入的回调函数。包含的钱包是基于在`executor.json`文件中配置的密钥

Reviewer: ```suggestion
- `execute`对象中目前有两个函数：`onEVM(options)`和`onSolana(options)`。这些函数需要一个Wormhole chain ID和一个有钱包对象注入的回调函数。包含的钱包是基于在`executor.json`文件中配置的密钥
```

         const result = await contract.processMyMessage(payload.vaa);
         this.logger.info(result);
       },
     });
   }
   else {   
     this.logger.error("Requested chainID is not an EVM chain, which is unsupported.");
   }
 }
```

在回调函数中，使用Ethers包创建一个[合约对象](https://docs.ethers.io/v5/api/contract/contract/#Contract){target=_blank}。其导入的ABI从`SimpleGeneralMessage`合约的编译中导出，所以该代码假设VAA中指定的消息接收者是或从`SimpleGeneralMessage`合约继承。

然后，该代码会尝试用VAA执行函数`processMyMessage(bytes32 VAA)`，这是此前被定义为用于中继消息的函数。该函数名字是为智能合约随意设置的，因为中继器可以指定调用任何函数。这说明了开发者可以更改该中继器代码！

```
     await execute.onEVM({
       chainId: destChainID,
       f: async (wallet, chainId) => {
         const contract = new ethers.Contract(recipient, abi, wallet.wallet);
         const result = await contract.processMyMessage(payload.vaa);
         this.logger.info(result);
       },
     });
```

最后一步就是检查`example-project/relayer-engine-config/common.json`。该配置文件控制着整个中继器的执行。请确保您在使用的TestNet EVM是列出在该文件中`supportedChains`对象内的。如果其未被列出，该插件不会正常运行。如果您在运行的一条链未被列出，您将需要以下列格式从[Wormhole的开发者文档](https://book.wormhole.com/reference/contracts.html#testnet){target=_blank}导入数据至配置文件。

> 该中继器还有许多其他配置。例如，`mode`字符串设置为`“BOTH”`以确保使用监听器和执行器插件，也可根据开发者需求选择只运行其中一个。此外，还有多个日志级别可供指定，如错误消息的`“error”`。然而，在本次演示中只需保留配置设置即可。

Reviewer: ```suggestion
该中继器还有许多其他配置。例如，`mode`字符串设置为`“BOTH”`以确保使用监听器和执行器插件，也可根据开发者需求选择只运行其中一个。此外，还有多个日志级别可供指定，如`“error”`可用来只记录错误消息。然而，在本次演示中只需保留配置设置即可。
```

```javascript
 "mode": "BOTH",
 "logLevel": "debug",
 ...
   {
     "chainId": 16,
     "chainName": "Moonbase Alpha",
     "nodeUrl": "https://rpc.api.moonbase.moonbeam.network",
     "bridgeAddress": "0xa5B7D85a8f27dd7907dc8FdC21FA5657D5E2F901",
     "tokenBridgeAddress": "0xbc976D4b9D57E57c3cA52e1Fd136C45FF7955A96"
   },
```

配置这样就行了！现在需要运行它。在您的终端实例（未运行间谍节点的实例），找到`example-project`文件夹。运行下列命令：

```
npm run start
```

您应该会在控制台中看到类似以下日志内容。

![Run the relayer](/images/builders/integrations/bridges/wormhole/wormhole-7.png)

### 通过Wormhole从Moonbase传送跨链消息 {: #send-message-from-moonbase }

现在，要想发送一条跨链消息，您只需要调用`sendMessage(string memory message, address destAddress, uint16 destChainId)`函数。

使用Remix接口。此范例将向Fantom测试网发送跨链消息，但您可以将`destChainId`替换成您想要的任何EVM。接着，检查以下事项：

1. 环境为**Injected Provider**，网络为1287（Moonbase Alpha）

> 2. 您的钱包里有来自[faucet](https://apps.moonbeam.network/moonbase-alpha/faucet/){target=_blank}的大量资金，以支付交易成本和目标链Gas包含的DEV

Reviewer: ```suggestion
2. 您的钱包里有来自[faucet](https://apps.moonbeam.network/moonbase-alpha/faucet/){target=_blank}的大量资金，以支付源链和目标链上的交易Gas成本
```

> 3. 在**sendMessage**调用的**message**输入中输入您选择的短消息（在本示例中为“this is a message”）

Reviewer: ```suggestion
3. 在**sendMessage**部分的**message**输入框中输入您选择的短消息（在本示例中为“this is a message”）
```

> 4. 将您的SimpleGeneralMessage实例的地址放在目标链的**destAddress**输入框中

Reviewer: ```suggestion
4. 将您的在目标链上的SimpleGeneralMessage实例的地址放入**destAddress**输入框中
```
